Use libnethost to find hostfxr

HaoK opened the pull request:
For https://github.com/dotnet/aspnetcore/issues/8838

Actual code changes are almost entirely in [src/Servers/IIS/AspNetCoreModuleV2/CommonLib/HostFxrResolver.cpp](https://github.com/dotnet/aspnetcore/pull/39810/files#diff-1a60a9f6e31eea0cdda80f76b0d862e51903028fb6a7ae823622b890c8e92d2c)

The majority of the rest of the changes are a lot of vcxproj cleanup deduping to make things sane, as well as force compilation/linking into release due to not having a debug version of libnethost.lib

There's now a set of shared build.settings files:

Build.Common.Settings (shared settings used by both dll and libs)
Build.Dll.Settings (the existing one, is meant for dlls, AspNetCore, InProc, OutOfProc)
Build.Lib.Settings (new one for static lib use, CommonLib, RequestHandler, gtest)
release.props (release specific settings, imported by build.common.settings)
debug.props (debug specific settings, imported by build.common.settings)
(I gave up on deduping IISLib for this first PR)

Side note: reenabling functional tests on local runs now since they all seem to be passing in reasonable time now:
StartupTests (90 ran in 22 minutes)

HaoK committed:
Building including checkpoint

HaoK committed:
try stuff

HaoK committed:
attempts

HaoK committed:
Tweaks

HaoK committed:
Revert some tweaks

HaoK committed:
Cleanup

HaoK committed:
Fix toolset versions to 142

HaoK committed:
Fixes

HaoK commented on src/Servers/IIS/AspNetCoreModuleV2/AspNetCore/AspNetCore.vcxproj:
```diff
@@ -123,12 +121,12 @@
       <PrecompiledHeaderFile>stdafx.h</PrecompiledHeaderFile>
       <PrecompiledHeaderOutputFile>$(IntDir)$(TargetName).pch</PrecompiledHeaderOutputFile>
-      <AdditionalIncludeDirectories>..\IISLib;.\Inc;..\CommonLib</AdditionalIncludeDirectories>
+      <AdditionalIncludeDirectories>..\IISLib;.\Inc;..\CommonLib;;C:\Github\aspnetcore\.dotnet\packs\Microsoft.NETCore.App.Host.win-x64\7.0.0-alpha.1.21612.5\runtimes\win-x64\native</AdditionalIncludeDirectories>
```
@dotnet/aspnet-build do you guys have any suggestions for what the proper way for me to reference into the packs\app.host would be here, I don't see anything that seems useful in the current msbuild variables.

Ideally I want something like `$(ReproRoot)\.dotnet\packs\Microsoft.NETCore.App.Host.win-$(Platform)\$(AppHostVersion)\runtimes\win-$(Platform)\native` so I can use it for x64/Win32/Arm64 without having to define this for each configuration

dougbu replied:
`$(MicrosoftNETCoreAppRuntimeVersion)` is not specific to Microsoft.NETCore.App versions or platforms; that should do for `$(AppHostVersion)`. Do _not_ use `$(MicrosoftNETCoreBrowserDebugHostTransportVersion)` because that's a package version and doesn't provide information about local paths when on servicing branches; very useful for runtime, host, and related package downloads but not for here.

`$(Platform)` should be fine from what I'm seeing. Well, as long as `$(Platform)` gets the same values as in our WiX and C# projects (`arm`, `arm64`, `x64`, `x86`). I'm not positive other than the `win-XYZ` bit being consistent in the two parts of the path.

HaoK replied:
Yeah I looked at AppRuntimeVersion already but that doesn't seem to matchup, here's what that results in today:

`LibNetHostPath = C:\Github\aspnetcore\\.dotnet\packs\Microsoft.NETCore.App.Host.win-x64\7.0.0-preview.2.22075.9\runtimes\win-x64\native`

But the version I get from restore seems to be: 22066.4

![image](https://user-images.githubusercontent.com/6537861/151478400-660e0be0-063d-4712-b8c5-2ac8e1c1dc7c.png)

dougbu replied:
Oh, I was wrong about alignment. The .NET SDK provides a few options that match that version. See e.g. .dotnet\sdk\7.0.100-alpha.1.22071.3\Microsoft.NETCoreSdk.BundledVersions.props in your repo. `$(AppHostPackVersion)` has a fairly on-point name for this case and might work.

HaoK replied:
Nice, so the last part is how do I reference that file, I see the sdk version in global.json but I don't see it hardcoded anywhere else as far as I can tell:

```
(aspnetcore) C:\Github\aspnetcore [haok/hostfxr ≡ +0 ~1 -0 !]> findstr /snip 22071.3 *.props
.dotnet\sdk\7.0.100-alpha.1.22071.3\Microsoft.NETCoreSdk.BundledVersions.props:24:    <NETCoreSdkVersion>7.0.100-alpha.1.22071.3</NETCoreSdkVersion>
```

Also I was just going to put this in the Directory.Build.Props for iis, but maybe we should put this in a shared props somewhere (maybe the RepoRoot props file?) since its probably useful 

```
  <Import Project="$(LocalDotNetRoot)\sdk\$(???)\Microsoft.NETCoreSdk.BundledVersions.props" />

```

dougbu replied:
Shouldn't need to import the file at all. Any project using the .NET SDK should have the `$(AppHostPackVersion)` property defined.

HaoK replied:
Are native projects somehow different then?  I wans't see any of the properties in the msbuild without explicitly importing

dougbu replied:
Yes, very different. Native projects don't use the .NET SDK.

You may need to add the property to eng/tools/GenerateFiles/Directory.Build.props.in and change the `$(_TemplateProperties)` property in GenerateFiles.csproj to pass `$(AppHostPackVersion)`. Something like

``` msbuild
<AppHostPackVersion>${AppHostPackVersion}</AppHostPackVersion>
```

and

``` msbuild
AppHostPackVersion = $(AppHostPackVersion);
```

The final file is imported into most projects in the repo, including native projects.

dougbu replied:
Oops, the bit for the template file should be conditional since we get it for free in most projects:

``` msbuild
<AppHostPackVersion Condition=" '$(AppHostPackVersion)' == '' ">${AppHostPackVersion}</AppHostPackVersion>
```

HaoK replied:
Awesome, looks like this did the trick, whitespace apparently matters in those template files, i.e. `AppHostPackVersion=$(AppHostPackVersion)` otherwise it generates it as ' 7.0.0xxx' with a leading space

HaoK committed:
Cleanup hostpack location

HaoK committed:
Cleanup unneeded include

HaoK committed:
Ignore linker warnings

HaoK committed:
Fix msbuild stuff

HaoK committed:
Make tests pass

HaoK committed:
Fix vcxproj

HaoK committed:
Undo some vxcproj changes

HaoK committed:
Undo more diffs

HaoK committed:
More diff cleanup

HaoK commented on src/Servers/IIS/IIS/src/Microsoft.AspNetCore.Server.IIS.csproj:
```diff
@@ -35,8 +35,8 @@
     <NativeProjectReference Include="$(MSBuildThisFileDirectory)..\..\AspNetCoreModuleV2\InProcessRequestHandler\InProcessRequestHandler.vcxproj" Platform="x64" />
-    <NativeProjectReference Include="$(MSBuildThisFileDirectory)..\..\AspNetCoreModuleV2\InProcessRequestHandler\InProcessRequestHandler.vcxproj" Platform="Win32" />
-    <NativeProjectReference Include="$(MSBuildThisFileDirectory)..\..\AspNetCoreModuleV2\InProcessRequestHandler\InProcessRequestHandler.vcxproj" Platform="ARM64" />    
+    <!--NativeProjectReference Include="$(MSBuildThisFileDirectory)..\..\AspNetCoreModuleV2\InProcessRequestHandler\InProcessRequestHandler.vcxproj" Platform="Win32" />
```
REVERT

HaoK committed:
Undo diffs

HaoK commented on src/Servers/IIS/AspNetCoreModuleV2/CommonLibTests/CommonLibTests.vcxproj:
```diff
@@ -230,15 +229,15 @@
       <GenerateDebugInformation>true</GenerateDebugInformation>
       <SubSystem>Console</SubSystem>
-      <AdditionalOptions>/NODEFAULTLIB:libucrt.lib /DEFAULTLIB:ucrt.lib %(AdditionalOptions)</AdditionalOptions>
+      <AdditionalOptions>/NODEFAULTLIB:libucrt.lib /DEFAULTLIB:ucrt.lib /ignore:4099 %(AdditionalOptions)</AdditionalOptions>
```
4099 is to ignore linker warning due to no pdbs from libnethost

elinor-fung replied:
These are the relevant issues for the lack of libnethost.pdb:
- https://github.com/dotnet/runtime/issues/63602
- https://github.com/dotnet/arcade/issues/8374

HaoK commented on src/Servers/IIS/AspNetCoreModuleV2/CommonLib/StringHelpers.cpp:
```diff
@@ -4,17 +4,6 @@
 #include "StringHelpers.h"
 #include "exceptions.h"
 
-bool ends_with(const std::wstring &source, const std::wstring &suffix, bool ignoreCase)
```
we now get an ends_with pulled in from libnethost

BrennanConroy replied:
Or namespace ours

dougbu replied:
Just curious: What goes South if you remove ours and use the libnethost one❔

HaoK commented on src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/InProcessRequestHandler.vcxproj:
```diff
@@ -72,7 +71,7 @@
     <UseDebugLibraries>false</UseDebugLibraries>
     <PlatformToolset>$(PlatformToolsetVersion)</PlatformToolset>
-    <WholeProgramOptimization>true</WholeProgramOptimization>
+    <WholeProgramOptimization>false</WholeProgramOptimization>
```
This optimization results in
```
LINK : fatal error LNK1000: Internal error during IMAGE::BuildImage [C:\Github\aspnetcore\src\Servers\IIS\AspNetCoreMod
uleV2\InProcessRequestHandler\InProcessRequestHandler.vcxproj]
```

dougbu replied:
Please open an issue about restoring optimization

HaoK committed:
Cleanup platformtoolsetversions

HaoK committed:
Centralize all the PlatformToolsetVersion

HaoK commented on src/Servers/IIS/AspNetCoreModuleV2/AspNetCore/AspNetCore.vcxproj:
```diff
@@ -42,8 +42,6 @@
     <LinkIncremental>false</LinkIncremental>
     <WindowsTargetPlatformVersion Condition=" '$(WindowsTargetPlatformVersion)' == '' ">10.0.18362.0</WindowsTargetPlatformVersion>
-    <PlatformToolsetVersion Condition=" '$(VisualStudioVersion)' == '17.0' ">v143</PlatformToolsetVersion>
-    <PlatformToolsetVersion Condition=" '$(PlatformToolsetVersion)' == '' ">v142</PlatformToolsetVersion>
```
@dougbu I believe you added these in https://github.com/dotnet/aspnetcore/commit/9217656f967cc6d040071df73b99cd94fddc5e84#diff-5f4ea5312e8f8d98f704201802678b117219b89f48cc8283330a652faf579749 now that we are using a static lib from runtime, we are required to use the same version they do, which is v142, so I've just removed all of these overrides since they break things, and hardcode v142 which is required in the directory.build props, do you see any problem with this?

HaoK replied:
I get that this was trying to give users some wiggle room previously, but now that we are consuming the libnethost.lib static library from the runtime repo, we are required to compile with the same toolset they are (v142).  

Actually this is going to cause another problem in the very near future, in that we require v143 tools for our arm64 builds, so @elinor-fung do we need runtime to upgrade to v143 tools, so we can consume libnethost on arm64?

dougbu replied:
Is the runtime repo doing CI builds exclusively on VS2019 agents❔ And, of so, how are they supporting VS1022 users❔ This isn't like other components, where the version can be controlled (at least to some extent) when choosing VS components.

HaoK replied:
I don't really have answers to those questions, but previously we were hardcoded to v142, and that's what we need to do for things to link against the static library currently.  At least in my VS today, after hardcoding v142, what I see in VS is that it just displays it as VS2019 next to the project

HaoK replied:
Looks like this for me and works fine in VS2022

![image](https://user-images.githubusercontent.com/6537861/152064796-197170a4-30eb-457b-8bdb-e074ab86a5d0.png)

HaoK committed:
Add comment

HaoK committed:
Fix Win32/Rename ends_with to avoid collision

HaoK committed:
Fix proj

HaoK committed:
Rename fx_ver again

HaoK committed:
Revert "Rename fx_ver again"

This reverts commit 85ae82b513bad1174c2532c55a209486f9ee3be0.

HaoK committed:
Undo some of the arm64 for now

HaoK committed:
Don't depend on arm64 for now

HaoK committed:
Skip ARM64 jobs, and use null params when no dotnet root

HaoK commented:
@elinor-fung yay progress, so looks like things are wired up now and working in some cases, I'm definitely seeing some large % of tests fail (sometimes non deterministically?), where get_hostfxr_path will fail, I'll see if I can isolate some locally with coretrace enabled, but here's an example of a helix failure where it looks like get_hostfxr failed from the inprocess dll, it worked from the shim(aspnetcorev2.dll), but failed given the same inputs in the aspnetcorev2_inprocess.dll: Do I need to do some fallback probing with different params if we don't find it the first time, i.e. in here? https://github.com/dotnet/aspnetcore/blob/481e0e995e68293460dc1696de7004a2f6d40216/src/Servers/IIS/AspNetCoreModuleV2/CommonLib/HostFxrResolver.cpp#L96

```
[aspnetcorev2.dll] Initializing logs for 'C:\h\w\AA410955\w\BBE809F9\e\ANCM\x64\aspnetcorev2.dll'. Process Id: 4160. File Version: 17.0.22034.0. Description: IIS ASP.NET Core Module V2. Commit: 18ab6eb8cb1ba6acf704557613bd184516c7fdbb.
[aspnetcorev2.dll] Resolving hostfxr parameters for application: 'C:\h\w\AA410955\p\dotnet-cli\dotnet.exe' arguments: '.\InProcessWebSite.dll ThrowInStartup' path: 'C:\h\w\AA410955\t\0009dbc2e1bd43e1b0d1c17b9907a73c\'
[aspnetcorev2.dll] Known dotnet.exe location: ''
[aspnetcorev2.dll] Process path 'C:\h\w\AA410955\p\dotnet-cli\dotnet.exe' is dotnet, treating application as portable
[aspnetcorev2.dll] Resolving absolute path to dotnet.exe from 'C:\h\w\AA410955\p\dotnet-cli\dotnet.exe'
[aspnetcorev2.dll] Found dotnet.exe at 'C:\h\w\AA410955\p\dotnet-cli\dotnet.exe'
[aspnetcorev2.dll] hostfxr.dotnet_root: 'C:\h\w\AA410955\p\dotnet-cli'
[aspnetcorev2.dll] hostfxr.assembly_path: 'C:\h\w\AA410955\t\0009dbc2e1bd43e1b0d1c17b9907a73c\'
[aspnetcorev2.dll] get_hostfxr_path 'C:\h\w\AA410955\p\dotnet-cli\host\fxr\7.0.0-preview.2.22078.1\hostfxr.dll'
[aspnetcorev2.dll] Resolving absolute path to hostfxr.dll from 'C:\h\w\AA410955\p\dotnet-cli\dotnet.exe'
[aspnetcorev2.dll] hostfxr.dll located at 'C:\h\w\AA410955\p\dotnet-cli\host\fxr\7.0.0-preview.2.22078.1\hostfxr.dll'
[aspnetcorev2.dll] Converted argument '.\InProcessWebSite.dll' to 'C:\h\w\AA410955\t\0009dbc2e1bd43e1b0d1c17b9907a73c\.\InProcessWebSite.dll'
[aspnetcorev2.dll] Parsed hostfxr options: dotnet location: 'C:\h\w\AA410955\p\dotnet-cli\dotnet.exe' hostfxr path: 'C:\h\w\AA410955\p\dotnet-cli\host\fxr\7.0.0-preview.2.22078.1\hostfxr.dll' arguments:
[aspnetcorev2.dll] Argument[0] = 'C:\h\w\AA410955\p\dotnet-cli\dotnet.exe'
[aspnetcorev2.dll] Argument[1] = 'C:\h\w\AA410955\t\0009dbc2e1bd43e1b0d1c17b9907a73c\.\InProcessWebSite.dll'
[aspnetcorev2.dll] Argument[2] = 'ThrowInStartup'
[aspnetcorev2.dll] Loading hostfxr from location C:\h\w\AA410955\p\dotnet-cli\host\fxr\7.0.0-preview.2.22078.1\hostfxr.dll
[aspnetcorev2.dll] Canceling standard stream pipe reader
[aspnetcorev2.dll] Loading request handler:  'C:\h\w\AA410955\t\0009dbc2e1bd43e1b0d1c17b9907a73c\x64\aspnetcorev2_inprocess.dll'
[aspnetcorev2.dll] Creating handler application
[aspnetcorev2_inprocess.dll] Initializing logs for 'C:\h\w\AA410955\t\0009dbc2e1bd43e1b0d1c17b9907a73c\x64\aspnetcorev2_inprocess.dll'. Process Id: 4160. File Version: 17.0.22034.0. Description: IIS ASP.NET Core Module V2 Request Handler. Commit: 18ab6eb8cb1ba6acf704557613bd184516c7fdbb.
[aspnetcorev2_inprocess.dll] Waiting for initialization
[aspnetcorev2_inprocess.dll] Starting app_offline monitoring in application 'C:\h\w\AA410955\t\0009dbc2e1bd43e1b0d1c17b9907a73c\'
[aspnetcorev2_inprocess.dll] Starting file watcher thread
[aspnetcorev2_inprocess.dll] Starting in-process worker thread
[aspnetcorev2_inprocess.dll] Resolving hostfxr parameters for application: 'C:\h\w\AA410955\p\dotnet-cli\dotnet.exe' arguments: '.\InProcessWebSite.dll ThrowInStartup' path: 'C:\h\w\AA410955\t\0009dbc2e1bd43e1b0d1c17b9907a73c\'
[aspnetcorev2_inprocess.dll] Known dotnet.exe location: 'C:\h\w\AA410955\p\dotnet-cli\dotnet.exe'
[aspnetcorev2_inprocess.dll] Process path 'C:\h\w\AA410955\p\dotnet-cli\dotnet.exe' is dotnet, treating application as portable
[aspnetcorev2_inprocess.dll] hostfxr.dotnet_root: 'C:\h\w\AA410955\p\dotnet-cli'
[aspnetcorev2_inprocess.dll] hostfxr.assembly_path: 'C:\h\w\AA410955\t\0009dbc2e1bd43e1b0d1c17b9907a73c\'
[aspnetcorev2_inprocess.dll] get_hostfxr_path failed
[aspnetcorev2_inprocess.dll] Event Log: 'Application 'C:\h\w\AA410955\t\0009dbc2e1bd43e1b0d1c17b9907a73c\' failed to start. Exception message:
[Information] Failed to find host fxr.' 

```

HaoK commented:
Hrm, this is really weird, so I'm seeing what looks like garbage in the corehosttrace on some of the failed runs, the logging in the test output looks fine, I ran the test 10 times via test explorer and it only failed 2 times with the mangled parameters.:

```
    | [1.921s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [2022-02-03T23:33:23.902Z, PID: 5096] [aspnetcorev2.dll] Initializing logs for 'C:\Github\aspnetcore\src\Servers\IIS\IIS\test\IIS.NewShim.FunctionalTests\bin\Release\net7.0\ANCM\x64\aspnetcorev2.dll'. Process Id: 5096. File Version: 17.0.22034.0. Description: IIS ASP.NET Core Module V2. Commit: 481e0e995e68293460dc1696de7004a2f6d40216.
| [1.921s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [2022-02-03T23:33:23.905Z, PID: 5096] [aspnetcorev2.dll] Resolving hostfxr parameters for application: 'C:\Github\aspnetcore\.dotnet\dotnet.exe' arguments: '.\InProcessWebSite.dll' path: 'C:\Users\haok\AppData\Local\Temp\27f8684e044a4966958fc2a5932ab893\'
| [1.921s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [2022-02-03T23:33:23.907Z, PID: 5096] [aspnetcorev2.dll] Known dotnet.exe location: ''
| [1.921s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [2022-02-03T23:33:23.908Z, PID: 5096] [aspnetcorev2.dll] Process path 'C:\Github\aspnetcore\.dotnet\dotnet.exe' is dotnet, treating application as portable
| [1.921s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [2022-02-03T23:33:23.909Z, PID: 5096] [aspnetcorev2.dll] hostfxr.dotnet_root: 'C:\Github\aspnetcore\.dotnet'
| [1.921s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [2022-02-03T23:33:23.910Z, PID: 5096] [aspnetcorev2.dll] hostfxr.assembly_path: 'C:\Users\haok\AppData\Local\Temp\27f8684e044a4966958fc2a5932ab893\'
| [1.921s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [2022-02-03T23:33:23.912Z, PID: 5096] [aspnetcorev2.dll] get_hostfxr_path failed, trying again with dotnet path as dotnet root.
| [1.921s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [2022-02-03T23:33:23.913Z, PID: 5096] [aspnetcorev2.dll] - hostfxr.dotnet_root: ''
| [1.921s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [2022-02-03T23:33:23.915Z, PID: 5096] [aspnetcorev2.dll] - hostfxr.assembly_path: 'C:\Users\haok\AppData\Local\Temp\27f8684e044a4966958fc2a5932ab893\'
| [1.921s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [2022-02-03T23:33:23.916Z, PID: 5096] [aspnetcorev2.dll] Event Log: 'Application 'C:\Users\haok\AppData\Local\Temp\27f8684e044a4966958fc2a5932ab893\' failed to start. Exception message:
| [1.921s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: Failed to find host fxr.' 
```

But when I look at the corhost.trace, I see what looks like the timestamps getting passed in:
```
Tracing enabled @ Thu Feb  3 23:33:23 2022 GMT
Using dotnet root parameter [2022-02-03T23:33:23.909Z] as runtime location.
A fatal error occurred. The folder [2022-02-03T23:33:23.909Z\host\fxr] does not exist
Using dotnet root parameter [] as runtime location.
A fatal error occurred. The folder [host\fxr] does not exist
```

HaoK committed:
Only use libnethost when we don't have dotnetExePath

HaoK committed:
Fix old code path

HaoK committed:
Make sure old code still works

HaoK committed:
throw when dotnet path is now different

HaoK commented:
/azp run

HaoK committed:
Fix build

HaoK committed:
Use old code as fallback

elinor-fung commented:
> LOG_INFOF(L"get_hostfxr_path failed"); 

In general, for diagnosing issues with the failed call, it would be useful to log out the return value. The APIs will return different error codes based on what failed, so that can help narrow down what is going on.

>  garbage in the corehosttrace 

I took a look through how the parameters were being set in [commit 481e0e9](https://github.com/dotnet/aspnetcore/commit/481e0e995e68293460dc1696de7004a2f6d40216). I suspect the issue is with the use of `parent_path()`:
https://github.com/dotnet/aspnetcore/blob/481e0e995e68293460dc1696de7004a2f6d40216/src/Servers/IIS/AspNetCoreModuleV2/CommonLib/HostFxrResolver.cpp#L81-L84

I'm not familiar with the `std::filesystem::path` APIs, but based on the [parent_path doc](https://en.cppreference.com/w/cpp/filesystem/path/parent_path), I'd guess what is happening is that the `path` returned by `parent_path()` has gone out of scope when `get_hostfxr_path` is called, so `params.dotnet_root` just ends up pointing to whatever happens to be at that address now (which would explain why it sometimes has what is expected and sometimes does not).

HaoK committed:
Pass dotnet path for second call

elinor-fung commented on src/Servers/IIS/AspNetCoreModuleV2/CommonLib/HostFxrResolver.cpp:
```diff
@@ -89,88 +89,81 @@ HostFxrResolver::GetHostFxrParameters(
+                get_hostfxr_parameters params;
                 params.size = sizeof(get_hostfxr_parameters);
                 params.assembly_path = applicationPhysicalPath.c_str();
                 params.dotnet_root = expandedProcessPath.parent_path().c_str();
```
I believe the return of `parent_path()` is going to be a [temporary object](https://en.cppreference.com/w/cpp/language/lifetime#Temporary_object_lifetime) here, so it will still be destructed before the `get_hostfxr_path` call (resulting in garbage in `params.dotnet_root`).

HaoK replied:
Okay going to show my C++ noob here, what's the proper pattern to do this call then, do I need to explicitly copy the string into another variable, like hostFxrPath for it not to get destroyed?

elinor-fung replied:
That is an option. I would just assign the return of `parent_path()`  to a variable such that you keep it around until you call `get_hostfxr_path`. For example:
```c++
{
    fs::path process_folder = expandedProcessPath.parent_path();
    params.dotnet_root = process_folder.c_str();
    result = get_hostfxr_path(...)
} // process_folder goes out of scope and is destructed - we don't need it anymore after get_hostfxr_path is done
```

elinor-fung replied:
If you are interested in seeing exactly what is going on, one of the things I usually do is just debug through a simple case in a console app. That would give you an easy way step though the difference between two cases like:
```c++
std::filesystem::path my_path = ...
{
    std::filesystem::path parent = my_path.parent_path();
    const wchar_t* ptr = parent.c_str();
    std::wcout << ptr << std::endl;
} // parent goes out of scope and is destructed
{
    const wchar_t* ptr = my_path.parent_path().c_str();
    // object returned by parent_path is immediately destructed after the call in the line above
    std::wcout << ptr << std::endl;
}
```
You can do something like look at the [Disassembly Window](https://docs.microsoft.com/visualstudio/debugger/how-to-use-the-disassembly-window) or set a [Function Breakpoint](https://docs.microsoft.com/visualstudio/debugger/using-breakpoints#BKMK_Set_a_breakpoint_in_a_source_file) for the destructor (`std::filesystem::path::~path`) to see when the path object is being destructed.

HaoK replied:
Thanks for the pointer, hopefully that this care of the non deterministic failures finally!

HaoK committed:
Fix parent_path issues

HaoK committed:
Introduce a helper

HaoK committed:
Cleanup

HaoK committed:
Cleanup

HaoK committed:
Add debug checks

HaoK committed:
Try arm64 again

HaoK committed:
Revert "Try arm64 again"

This reverts commit c29843370b6046a1b1c80a479f0eb4df9d573ea4.

HaoK committed:
Try some vctools fixes

HaoK committed:
Revert "Try some vctools fixes"

This reverts commit a228edfb703085d58ae09f879fea5bfd4bdb0938.

HaoK committed:
Fix VCToolsVersion to 14.29.30133

HaoK committed:
Revert "Revert "Try arm64 again""

This reverts commit 1213d61a2c64d5f74c2f951e8eb5daafb4189ad4.

HaoK committed:
Fix link

HaoK committed:
Add back arm64 assets

HaoK committed:
Restore arm64 build stuff

HaoK committed:
Remove some code

HaoK committed:
Remove SkipnonHelix

HaoK committed:
Remove old code and debugging guards

Reenable some global version tests

working local Debug stuff

Remove launch settings

Revert "working local Debug stuff"

This reverts commit e61d0d9770540e16846a3bba0fcd53e203092f58.

Cleanup vcxproj

HaoK commented:
@elinor-fung after chatting with @adityamandaleeka today we came to the conclusion that not having a debug version of libnethost.lib is going to be blocking, do you think in the short term you could just manually build x86/x64/arm32 versions libnethostd.lib that we can use until the real thing is ready?  We could just check those in for now for our debug builds.

The other ask we had was if it would be possible for the path probing of dotnet logic to be moved into get_host_fxr_path itself, while we are able to remove some of our dotnet folder lookup logic, we'd also like to remove the code we have that is calling where and path probing to find dotnet, what do you think about making this a responsibility of nethost to attempt to find dotnet on the path if it isn't given an absolute path to it?  Similar to our logic today, maybe if dotnet_root is set to "dotnet.exe" or "dotnet", that could be the mechanism to request path probing for the dotnet_root?

HaoK commented:
Specifically we'd like behavior that mimic'd:

https://github.com/dotnet/aspnetcore/blob/main/src/Servers/IIS/AspNetCoreModuleV2/CommonLib/HostFxrResolver.cpp#L223 so we didn't have to find dotnet_root ourselves

elinor-fung commented:
> not having a debug version of libnethost.lib is going to be blocking

Is that just from not wanting to use the release CRT in your debug builds? Do you need the debug CRT for some non-production innerloop/dev scenario? Manual build + check in sounds pretty sketchy.

> calling where and path probing to find dotnet, what do you think about making this a responsibility of nethost to attempt to find dotnet on the path

That's really not the intent of `get_hostfxr_path`. Particularly for 'calling where', the host has intentionally stayed out of that kind of probing business and left it up to the OS. I know there were involved discussions about looking at `PATH` and opting not to do it, but I can't seem to find a good link (from before I was on the team) - here is a more recent, short summary: https://github.com/dotnet/runtime/issues/60912#issuecomment-952874054.

The second portion (after `InvokeWhereToFindDotnet`) of the code you linked (looking up a registered installation, looking at a default install path, etc.) is what `nethost` is meant to do when you *don't* pass in `dotnet_root` - passing in `dotnet_root` means 'I know exactly where dotnet root is - use this value'.

HaoK commented:
@dotnet/aspnet-build can you take a quick look build related changes so far, the main changes to review is locking the native toolsetversion to 14.29.30133 for the ANCM/native IIS projects since they are required, and the libnethost path msbuild stuff in [src/Servers/IIS/Directory.Build.props](https://github.com/dotnet/aspnetcore/pull/39810/files#diff-98ed2c85a393e5fd8875c78597de9e8737f2fdbbffb90f6c94d086cfd5a8c31a)

dougbu commented on eng/tools/GenerateFiles/Directory.Build.props.in:
```diff
@@ -2,5 +2,6 @@
   <PropertyGroup>
     <DefaultNetCoreTargetFramework>${DefaultNetCoreTargetFramework}</DefaultNetCoreTargetFramework>
     <TreatWarningsAsErrors Condition="'$(BuildingInsideVisualStudio)' != 'true'">true</TreatWarningsAsErrors>
+    <AppHostPackVersion Condition=" '$(AppHostPackVersion)' == '' ">${AppHostPackVersion}</AppHostPackVersion>
```
Please remove this and the change in GenerateFiles.csproj. See #39965

dougbu replied:
Ah, I'm starting to get this based on your comments in src/Servers/IIS/Directory.Build.props. Suspect I made the suggestion and got confused by the property name. Let's at least change the `$(AppHostPackVersion)` property name so that it doesn't sound like something we're overriding in the .NET SDK. Make it visibly unique to avoid similar confusion in the future.

A comment like "for `$(LibNetHostPath)` setting in src/Servers/IIS/Diretory.Build.props" would also help.

HaoK replied:
Right this was your suggestion to get things to flow for native projects, so what do you want this name to be?

HaoK replied:
Could go with LibNetHostAppPackVersion?

dougbu replied:
> Could go with LibNetHostAppPackVersion?

Works for me

wtgodbe commented on src/Servers/IIS/Directory.Build.props:
```diff
@@ -1,6 +1,13 @@
+  
   <PropertyGroup>
     <KestrelSharedSourceRoot>$(MSBuildThisFileDirectory)..\Kestrel\shared\</KestrelSharedSourceRoot>
+    <HostArchDir>$(Platform)</HostArchDir>
```
nit, but I prefer `HostArch` to `HostArchDir`

HaoK replied:
Looks better to me too, will go with that

wtgodbe commented on src/Servers/IIS/Directory.Build.props:
```diff
@@ -1,6 +1,13 @@
+    <HostArchDir>$(Platform)</HostArchDir>
+    <HostArchDir Condition="'$(Platform)' == 'Win32'">x86</HostArchDir>
+    <HostArchDir Condition="'$(Platform)' == 'AnyCPU'">x64</HostArchDir>
+    <LibNetHostPath>$(NetCoreTargetingPackRoot)\Microsoft.NETCore.App.Host.win-$(HostArchDir)\$(AppHostPackVersion)\runtimes\win-$(HostArchDir)\native</LibNetHostPath>
```
Any chance this could ever change?

HaoK replied:
This is meant to change 'automatically' with every SDK update, that stuff comes from the Directory.props.in magic

HaoK replied:
See the stuff in GenerateProps: [eng/tools/GenerateFiles/GenerateFiles.csproj](https://github.com/dotnet/aspnetcore/pull/39810/files#diff-079db1ca47ec56a5adb02a25bab9b57e1b359f8872eb6d3aea53103b962f84e7)

wtgodbe replied:
What I mean is that you've hard coded the format of the path to libnethost here, what happens if they start dropping that into a different location?

HaoK replied:
We'd have to update this path whenever we take SDK update, I'd imagine this wouldn't change often as this is the dotnet SDK runtime packs layout

HaoK replied:
Our builds would fail immediately since we use this path to link against libnethost.lib which no longer would be found so its not something we can miss

dougbu commented on src/Servers/IIS/AspNetCoreModuleV2/CommonLibTests/CommonLibTests.vcxproj:
```diff
@@ -40,7 +40,6 @@
     <Keyword>Win32Proj</Keyword>
     <WindowsTargetPlatformVersion Condition=" '$(WindowsTargetPlatformVersion)' == '' ">10.0.18362.0</WindowsTargetPlatformVersion>
-    <PlatformToolsetVersion Condition=" '$(VisualStudioVersion)' == '17.0' ">v143</PlatformToolsetVersion>
     <PlatformToolsetVersion Condition=" '$(PlatformToolsetVersion)' == '' ">v142</PlatformToolsetVersion>
```
Why is this still here i.e. can we please set `$(PlatformToolsetVersion)` in just one place❔

HaoK replied:
Test vcxproj have their own quirks, I'm not going to dedupe this (or IISLib cs it has different warning levels) for this PR, but the rest of these are set once now in the shared Build.*.Settings stuff

dougbu commented on src/Servers/IIS/AspNetCoreModuleV2/OutOfProcessRequestHandler/OutOfProcessRequestHandler.vcxproj:
```diff
@@ -41,13 +41,11 @@
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.Default.props" />
   <PropertyGroup Label="Configuration">
-    <VCToolsVersion />
+    <VCToolsVersion>14.29.30133</VCToolsVersion>
```
I _think_ this property is set to a consistent value. Can that be done in Build.settings❔

dougbu commented on src/Servers/IIS/Directory.Build.props:
```diff
@@ -1,6 +1,13 @@
+    <HostArchDir Condition="'$(Platform)' == 'AnyCPU'">x64</HostArchDir>
+    <LibNetHostPath>$(NetCoreTargetingPackRoot)\Microsoft.NETCore.App.Host.win-$(HostArchDir)\$(AppHostPackVersion)\runtimes\win-$(HostArchDir)\native</LibNetHostPath>
+    <!-- This tools version MUST match the same version as runtime uses to compile libnethost.lib -->
+    <PlatformToolsetVersion>v142</PlatformToolsetVersion>
```
I'm surprised this works for users of VS2022 but saw the earlier comments about both toolsets being available in that release. Have you tested building w/ the native build grabbing `msbuild` from that release to be sure❔

HaoK replied:
I only build with VS2022, so we are fine in that scenario, I don't think we will be able to support people who don't have VS2022 after this PR goes in, we probably will have to update our requirements to be the same as what runtime does here https://github.com/dotnet/runtime/blob/main/docs/workflow/requirements/windows-requirements.md#visual-studio

dougbu replied:
We build from the command line on VS2019 in the CI. You haven't broken that.

dougbu commented on src/Servers/IIS/IIS/test/IIS.NewShim.FunctionalTests/Properties/launchSettings.json:
```diff
@@ -0,0 +1,8 @@
+{
```
Why is this file being checked in❔

dougbu reviewed (changes requested):
Oops, meant to request changes

adityamandaleeka commented on src/Servers/IIS/AspNetCoreModuleV2/AspNetCore/AspNetCore.vcxproj:
```diff
@@ -144,9 +142,10 @@
       <SubSystem>Windows</SubSystem>
       <GenerateDebugInformation>true</GenerateDebugInformation>
-      <AdditionalDependencies>kernel32.lib;user32.lib;advapi32.lib;shell32.lib;ole32.lib;oleaut32.lib;uuid.lib;ahadmin.lib;ws2_32.lib;iphlpapi.lib;version.lib</AdditionalDependencies>
+      <AdditionalDependencies>kernel32.lib;user32.lib;advapi32.lib;shell32.lib;ole32.lib;oleaut32.lib;uuid.lib;ahadmin.lib;ws2_32.lib;iphlpapi.lib;version.lib;$(LibNetHostPath)\libnethost.lib</AdditionalDependencies>
```
Is the $(LibNetHostPath) needed here if it's already in the AdditionalLibraryDirectories?

adityamandaleeka replied:
(and same elsewhere)

HaoK replied:
Preference between just using the absolute lib path instead of setting the Directory?  I think I'd rather just nuke the additional library directory changes personally

adityamandaleeka commented on src/Servers/IIS/AspNetCoreModuleV2/CommonLibTests/CommonLibTests.vcxproj:
```diff
@@ -50,7 +49,7 @@
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.Default.props" />
   <PropertyGroup Label="Configuration">
-    <VCToolsVersion />
+    <VCToolsVersion>14.29.30133</VCToolsVersion>
```
Why are we adding these?

HaoK replied:
The default we get from VS2022 is higher, its 14.30xx, so we need to go lower than current because we have to use the v142 toolset that matches runtimes, I can't remember which variant of the build it was that would fail without this, I think its the arm64 or x86, this wasn't necessary for x64, I added this later to get the other variants working

adityamandaleeka commented on src/Servers/IIS/build/Build.Settings:
```diff
@@ -4,8 +4,7 @@
      <Platform Condition="'$(Platform)' == ''">Win32</Platform>
-     <PlatformToolsetVersion Condition=" '$(VisualStudioVersion)' == '17.0' ">v143</PlatformToolsetVersion>
-     <PlatformToolsetVersion Condition=" '$(PlatformToolsetVersion)' == '' ">v142</PlatformToolsetVersion>
+     <PlatformToolsetVersion>v142</PlatformToolsetVersion>
```
Do we need this in both Directory.build.props and here?

HaoK replied:
I still need to clean up some of the projects, the libs aren't including build.settings, only the dlls are, I'll try to normalize some of these in the next iteration

HaoK committed:
Working checkpoint

Remove hardcode

HaoK commented on src/Servers/IIS/AspNetCoreModuleV2/CommonLib/CommonLib.vcxproj:
```diff
@@ -195,7 +183,9 @@
     </Link>
     <Lib>
-      <AdditionalLibraryDirectories>..\iislib</AdditionalLibraryDirectories>
+      <AdditionalLibraryDirectories>..\iislib;C:\Github\aspnetcore\.dotnet\packs\Microsoft.NETCore.App.Host.win-x64\7.0.0-alpha.1.21612.5\runtimes\win-x64\native</AdditionalLibraryDirectories>
```
Oops yeah missed one of these, thanks!

HaoK committed:
PR feedback

HaoK committed:
Working vcxproj clean up

Update aspnetcore.vcxproj

Cleanup InProc vcxproj

More dedupe

Dedupe out of proc proj

Nuke more dupes

Move stuff into shared Build.Common.Settings

More cleanup

More vcxproj surgery

More build cleanup

Move Win32Proj to Common

Cleanup gtest

BrennanConroy commented on eng/tools/GenerateFiles/Directory.Build.props.in:
```diff
@@ -2,5 +2,6 @@
   <PropertyGroup>
     <DefaultNetCoreTargetFramework>${DefaultNetCoreTargetFramework}</DefaultNetCoreTargetFramework>
     <TreatWarningsAsErrors Condition="'$(BuildingInsideVisualStudio)' != 'true'">true</TreatWarningsAsErrors>
+    <LibNetHostAppPackVersion Condition=" '$(LibNetHostAppPackVersion)' == '' ">${LibNetHostAppPackVersion}</LibNetHostAppPackVersion>
```
If variable not set, set it to the not set variable?

HaoK replied:
Maybe this line isn't needed at all @dougbu ?  Since we set it in eng/tools/GenerateFiles/GenerateFiles.csproj explicitly?

HaoK replied:
For a current datapoint: after a restore.cmd, this is what is in artifacts\bin\GenerateFiles\Directory.Build.props:
```
<Project>
  <PropertyGroup>
    <DefaultNetCoreTargetFramework>net7.0</DefaultNetCoreTargetFramework>
    <TreatWarningsAsErrors Condition="'$(BuildingInsideVisualStudio)' != 'true'">true</TreatWarningsAsErrors>
    <LibNetHostAppPackVersion Condition=" '$(LibNetHostAppPackVersion)' == '' ">7.0.0-preview.2.22103.2</LibNetHostAppPackVersion>
  </PropertyGroup>
</Project>
```

HaoK replied:
So I think this change is correct, as its generating a Directory.Build.props.in and hardcoding the version, as a default if its not overriden in the actual vcxproj.

dougbu replied:
This is a .props file i.e. it's loaded before your project. You might as well remove the `Condition` and override the value in a project file unconditionally (if you need to).

dougbu replied:
> Maybe this line isn't needed at all @dougbu ?

To clarify: The line sets the property to the value that GenerateFiles.csproj provides only if the property isn't already set. Unless the property is set by something the .vcxproj imports fairly early on, the `Condition` should never be `true`. On the other hand, the .vcxproj files manually import the root Directory.Build.props and **do** have a chance to set a value before that import.

So, it's up to you how you handle this property. Definitely don't bother w/ the `Condition` if projects never need to override it and the core C++ stuff doesn't provide a value.

dougbu replied:
nit: I don't see any overrides. Suggest removing the `Condition` here.

BrennanConroy commented on src/Servers/IIS/AspNetCoreModuleV2/CommonLib/fx_ver.cpp:
```diff
@@ -7,26 +7,26 @@
 static bool validIdentifiers(const std::wstring& ids);
 
-size_t index_of_non_numeric(const std::wstring& str, size_t i)
+size_t index_of_non_numeric2(const std::wstring& str, size_t i)
```
Why

HaoK replied:
Ah because we copied this file and we get a name collision, I forgot about these, I'll try your namespace suggestion now to see if we can undo the renames

BrennanConroy replied:
Does that mean we can just use the runtimes version instead of rolling a copy?

HaoK replied:
Probably, what does that look like in practice, defining our headers as externs or something? and just deleting the cpp file?

HaoK replied:
Looks like I can just delete our fx_ver.cpp, and I just need to link our CommonLibTests.vxcproj against libnethost for this to work... one more time into mucking around with vcxproj!

HaoK replied:
Added removing fx_ver to the meta issue https://github.com/dotnet/aspnetcore/issues/39281 to take up later, for some reason there's a linker error only in Win32/x86 without our implementation of fx_ver, so nuking fx_ver.cpp will be left to a future PR.  But at least we have a partially cleaned up CommonLibTests now for this PR

BrennanConroy commented on src/Servers/IIS/AspNetCoreModuleV2/CommonLib/HostFxrResolver.cpp:
```diff
@@ -56,12 +59,43 @@ HostFxrResolver::GetHostFxrParameters(
         {
-            dotnetExePath = GetAbsolutePathToDotnet(applicationPhysicalPath, expandedProcessPath);
+            // need to find dotnet for get_host_fxr_path when dotnet is launched from the path
+            if (!equals_ignore_case(expandedProcessPath, L"dotnet.exe")) {
```
Put `{`'s on their own line

BrennanConroy commented on src/Servers/IIS/AspNetCoreModuleV2/IISLib/IISLib.vcxproj:
```diff
@@ -40,22 +40,20 @@
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'" Label="Configuration">
     <ConfigurationType>StaticLibrary</ConfigurationType>
-    <UseDebugLibraries>true</UseDebugLibraries>
+    <UseDebugLibraries>false</UseDebugLibraries>
```
Tracked somewhere so we can revert eventually?

HaoK replied:
Depends on whether we can convince runtime to produce a debug version of libnethost, but yeah good idea I'll file an issue tracking this

HaoK replied:
Added it to the meta issue https://github.com/dotnet/aspnetcore/issues/39281

HaoK committed:
Revert fx_ver rename and use aspnet namespace

HaoK committed:
PR feedback

Minor cleanup

More Cleanup/CommonLibTests

Move more settings into Common.Settings

More build cleanup

Revert stuff

Revert "Revert stuff"

This reverts commit 7faa512f84631f7ef7588e498c6a70c62c8ea350.

Undo -nobuild changes to iis tests to fix running tests locally

Revert "Move some targets into a Directory.Build.targets file (#39712)"

This reverts commit 209ff13647b57d16f9ed48c9e83c54277ff106fb.

Bring back directory build

Revert "Revert fx_ver rename and use aspnet namespace"

This reverts commit d9c3d281061a355870a708a97f81fde2022dd519.

Revert "Revert "Revert fx_ver rename and use aspnet namespace""

This reverts commit e2b34b88cde041b9e4bf1137eb4c07b9788b2232.

Revert "Bring back directory build"

This reverts commit ad9007e3ef6b7fce1c3677aa8cfa891c23fea32e.

Revert "Revert "Move some targets into a Directory.Build.targets file (#39712)""

This reverts commit e7937bc84443a4f9810bb3af70ebb8f080617737.

Unskip some more local tests

HaoK commented:
/azp run

dougbu commented on src/Servers/IIS/AspNetCoreModuleV2/AspNetCore/AspNetCore.vcxproj:
```diff
@@ -1,237 +1,36 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
-  <ItemGroup Label="ProjectConfigurations">
```
Nice❕ Is this file even 20% its previous size❔

HaoK replied:
317 lines to 116, so like 33% but its much saner to make changes to now, it was insane how many copies of compiler flags there were before

dougbu commented on src/Servers/IIS/AspNetCoreModuleV2/CommonLibTests/CommonLibTests.vcxproj:
```diff
@@ -2,7 +2,8 @@
   </PropertyGroup>
-  <ItemGroup Label="ProjectConfigurations">
+
+ <ItemGroup Label="ProjectConfigurations">
```
nit: Misaligned indentation❔

HaoK replied:
Is there a common tab pattern we are supposed to use for msbuild stuff?  I noticed the settings spacing was slightly different than the vcxproj, I just tried to keep the spacing the same in each file rather than normalizing

dougbu replied:
We generally follow regular XML nesting w/ two spaces per indentation level and a fair amount of randomness for attribute indentation if lines get too long

dougbu commented on src/Servers/IIS/build/Build.Common.Settings:
```diff
@@ -0,0 +1,122 @@
+   </PropertyGroup>
+
+   <PropertyGroup Condition="'$(Configuration)' == 'Release'">
+     <UseDebugLibraries>false</UseDebugLibraries>
```
Just curious but why would we set this `true` for a `Release` build❔

dougbu commented on src/Servers/IIS/build/Build.Common.Settings:
```diff
@@ -0,0 +1,122 @@
+<?xml version="1.0" encoding="utf-8"?>
+ <Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <ItemGroup Label="ProjectConfigurations">
```
Why not import src/Servers/IIS/build/Config.Definitions.Props here❔

HaoK replied:
Great idea, I didn't even notice that existed, will add the arm64 to that and just import that, thanks for pointing that out

HaoK replied:
So looks like the only thing missing from Config.Definitions.Props is the Any CPU configuration, is this something we need or can I just drop Any CPU from all of our projects as part of this cleanup, this isn't something we ship anyways right?

dougbu replied:
We're talking about what VS expects here and I'm not sure what will happen w/o AnyCPU in the options.

dougbu commented on src/Servers/IIS/build/Build.Common.Settings:
```diff
@@ -0,0 +1,122 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Suggest adding comments at the top of these files indicating which build/*.Setting or *.props file should be imported when

dougbu reviewed (commented):
nits and suggestions mostly&hellip;

HaoK commented:
/azp run

HaoK commented:
@dougbu it looks like this change with the no build is causing the issues (https://github.com/dotnet/aspnetcore/commit/633b3d7f459aa20067ddd088d9106fcdfb0c7bbb), I'm reverting the IIS functional test changes as part of this PR since there was no easy way for me to get the tests to work again until I reverted this change and it caused all iis tests to fail publishing locally until I reverted it because the publish directories like `C:\Github\aspnetcore\src\Servers\IIS\IIS\test\testassets\InProcessWebSite\bin\Release\net7.0\publish\InProcessWebSite-Portable` were missing and nothing seemed to create them until I reverted the changes

dougbu commented:
@HaoK we need another way forward that doesn't undo 633b3d7f459a. Let's chat sometime tomorrow. I'll look at what's in this branch in the meantime.

HaoK commented:
Its not anything specific to this branch, the tests don't work in main either, I have to use a branch against main as well to debug the differences in the failing tests, 

IIS functional tests are just broken right now in main, I'm not reverting the entire commit, just the handful of changes in the iis directories, see https://github.com/dotnet/aspnetcore/commit/1c81661a805c747c92bf23d6cdf1d7f4513576bd

dougbu commented:
Reverting that part of the commit is likely what broke the Helix tests. In any case, let's chat tomorrow about a way to unbreak local tests which moves us forward and not back. I probably missed something when testing IIS locally as part of that work; let's figure out what's missing.

HaoK commented:
No, I actually have a bug in my code, the current stdout redirect errors in this PR are real, this PR is unwieldy enough, I'm happy to help you isolate what was broken in the local test flow in a follow up PR, but I'm not going to add more build changes on top.  

If you prefer, I can file a PR that undos the iis test changes immediately in main to decouple unblocking running IIS tests locally apart from this PR, but since that isn't blocking anyone but me, I'm happy to keep it in my branch.  Making it impossible to run IIS functional tests locally for a little helix build time gain doesn't seem like an acceptable tradeoff

dougbu commented:
@HaoK sorry for the bother. Please revert https://github.com/dotnet/aspnetcore/pull/39810/commits/0f5e802efddf11149fed9db118badd8b76114b81 and either wait for me to get #40379 in or `cherry-pick` https://github.com/dotnet/aspnetcore/pull/40379/commits/bd800e43118e827febe57e71a0f2f5711818c5df instead.

HaoK commented:
Finally isolated what's going on with the failing tests but still not sure why.  The set of tests that are failing are trying to exercise corehost tracing, and expecting the tracing to show up in stderr, for some reason this is no longer showing up.  If I set corehost_tracefile for the test I do see it tracing, so there must be some slight difference causing corehost tracing not to get dumped into std error.

HaoK committed:
Disable fastuptodate check

Revert "Undo -nobuild changes to iis tests to fix running tests locally"

This reverts commit 4d6f106428851162d88dc24a6275737a18349a26.

HaoK committed:
Fix tests

This reverts commit 4d6f106428851162d88dc24a6275737a18349a26.

HaoK committed:
Add debug/release props

adityamandaleeka commented on src/Servers/IIS/AspNetCoreModuleV2/CommonLib/HostFxrResolver.cpp:
```diff
@@ -12,6 +12,9 @@
 #include "StringHelpers.h"
 #include "RegistryKey.h"
 
+#define NETHOST_USE_AS_STATIC
```
Please add a comment here to explain what this does.

adityamandaleeka commented on src/Servers/IIS/AspNetCoreModuleV2/CommonLib/HostFxrResolver.h:
```diff
@@ -50,22 +50,17 @@ class HostFxrResolver
-    FindHighestDotNetVersion(
-        std::vector<std::wstring> & vFolders
-    );
+        BOOL
```
I'd just use bool here... no reason for this to be BOOL (and same with the one above).

Please fix the indentation too.

adityamandaleeka commented on src/Servers/IIS/AspNetCoreModuleV2/CommonLib/HostFxrResolver.cpp:
```diff
@@ -144,6 +186,34 @@ HostFxrResolver::GetHostFxrParameters(
+    hostfxrPath.resize(size);
+
+    get_hostfxr_parameters params;
+    params.size = sizeof(get_hostfxr_parameters);
```
Is info the right level for this?

HaoK replied:
I don't think the info level matters much for native code, basically the only way you get anything is to turn on ancm debug logs, which will write this to the file, the logging is pretty verbose as this is pretty much all we are going to get later on if people have issues with this code, so I think its important we get the parameters here, otherwise we will be left with the extremely verbose corehost tracing only

HaoK committed:
Revert "Fix tests"

This reverts commit 53cc3a35982c6ccd735b30d320d4ab5041c201d0.

adityamandaleeka commented on src/Servers/IIS/build/debug.props:
```diff
@@ -0,0 +1,27 @@
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Thank you for doing all of this build cleanup!

adityamandaleeka commented on src/Servers/IIS/AspNetCoreModuleV2/CommonLib/HostFxrResolver.cpp:
```diff
@@ -121,7 +158,12 @@ HostFxrResolver::GetHostFxrParameters(
+                fs::path dotnetExePathParent = dotnetExePath.parent_path();
+                if (!TryGetHostFxrPath(hostFxrDllPath, dotnetExePathParent, applicationPhysicalPath))
+                {
+                    throw InvalidOperationException(format(L"get_hostfxr_path failed"));
```
Does this exception get surfaced to the user? Might be worth plumbing through the return value from get_hostfxr_path?

adityamandaleeka replied:
Also, just a thought: do we want to tell people how to collect the traces from nethost when stuff goes wrong?

HaoK replied:
That's actually the last set of tests that are broken right now, CoreHostTrace is the only way you can see what's going wrong since, we already log the return value in the logs, but I can return an HR instead of a bool so we have the code

adityamandaleeka commented on src/Servers/IIS/AspNetCoreModuleV2/CommonLib/HostFxrResolver.cpp:
```diff
@@ -56,12 +59,46 @@ HostFxrResolver::GetHostFxrParameters(
         if (dotnetExePath.empty())
         {
-            dotnetExePath = GetAbsolutePathToDotnet(applicationPhysicalPath, expandedProcessPath);
+            // need to find dotnet for get_host_fxr_path when dotnet is launched from the path
```
I'm not sure why we care about this scenario... would love to know if anyone uses it. We could get rid of a lot of complexity in our code if we don't have to worry about loading dotnet when it's not properly installed/registered.

adityamandaleeka reviewed (approved):
I left some comments, but overall this looks good. Thanks again for cleaning up the build! Given that it's such a large diff, I'm sure there are things we'll discover after merge, so I'm looking forward to getting it in and getting some mileage on it.

HaoK commented:
@dougbu anything else you still want addressed in this PR for the build side of things?  I just need to hunt down this last functional regression but I want to make sure everything else is good to go

dougbu commented on src/Servers/IIS/AspNetCoreModuleV2/OutOfProcessRequestHandler/OutOfProcessRequestHandler.vcxproj:
```diff
@@ -1,235 +1,24 @@
-      <ModuleDefinitionFile>Source.def</ModuleDefinitionFile>
-      <AdditionalDependencies>kernel32.lib;user32.lib;advapi32.lib;shell32.lib;ole32.lib;oleaut32.lib;uuid.lib;ahadmin.lib;ws2_32.lib;iphlpapi.lib;version.lib;Rpcrt4.lib;winhttp.lib</AdditionalDependencies>
-      <LinkTimeCodeGeneration>UseLinkTimeCodeGeneration</LinkTimeCodeGeneration>
+<AdditionalDependencies>kernel32.lib;user32.lib;advapi32.lib;shell32.lib;ole32.lib;oleaut32.lib;uuid.lib;ahadmin.lib;ws2_32.lib;iphlpapi.lib;version.lib;Rpcrt4.lib;winhttp.lib</AdditionalDependencies>
```
nit: This element should be nested deeper
```suggestion
 
 
 
 <AdditionalDependencies>kernel32.lib;user32.lib;advapi32.lib;shell32.lib;ole32.lib;oleaut32.lib;uuid.lib;ahadmin.lib;ws2_32.lib;iphlpapi.lib;version.lib;Rpcrt4.lib;winhttp.lib</AdditionalDependencies>
```

dougbu commented on src/Servers/IIS/build/release.props:
```diff
@@ -0,0 +1,51 @@
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+
+<!--
```
nit: You have a weird XML editor that doesn't understand indentation.

dougbu reviewed (approved):
Nothing objectionable @HaoK, just nit-picking and asking curiosity questions

HaoK commented:
> Just curious: What goes South if you remove ours and use the libnethost one❔

I got some compiler errors, its possible its doable, I'll add it to the main cleanup tracking issue to consider eventually, but theirs is typed like:

```
bool ends_with(const pal::string_t& value, const pal::string_t& suffix, bool match_case)
```
https://github.com/dotnet/runtime/blob/df69293bdbe75bf5903583c472fcd257119dbda4/src/native/corehost/hostmisc/utils.cpp#L32

vs ours in addition to being implemented differently

```
bool endsWith(const cstd::wstring& source, const std::wstring& suffix, bool ignoreCase
```

HaoK committed:
PR Feedback

adityamandaleeka commented on src/Servers/IIS/AspNetCoreModuleV2/CommonLib/HostFxrResolver.cpp:
```diff
@@ -186,7 +187,7 @@ HostFxrResolver::GetHostFxrParameters(
 }
 
-BOOL
+bool
```
nit: change the TRUE/FALSE to true/false when changing BOOL to bool.

HaoK committed:
Verify inprocess proj is issue

HaoK committed:
Restore more IIS tests

HaoK committed:
Revert "Verify inprocess proj is issue"

This reverts commit 8e933e5963760fb5db57b735c559f7a2e098e2cb.

HaoK committed:
Add linker nodefaultlib to dll settings

HaoK committed:
Use true/false

HaoK committed:
Fix comment
